joomla/joomla-cms
# [ACL] Don't explicit add the parent asset permissions on creating a new item + other issues

Pull Request for ACL Issues (part 2).
#### Summary of Changes

What this PR does:
- Stop explicit writing the parent asset permissions on creating a new item (module, article, etc). Now all are created with Inhrited in all groups/actions so they Inhrited them.
- Displays the correct calculated permissions in item <-> section(s) <-> component <-> global config scenario.
- On create a new item (module, article, etc) and not saving the Permmission tab calculated are now the component ones.
- Further simplifies the ACL rules code to make it more undertstandable.
- Solves some other minor bugs.

To get the explicit item ACL permission in a item <-> section(s) <-> component <-> global config scenario i had to change JAccess::getAssetRules() behaviour.

**Mantainers** before merging please confirm this is correct and will not have side effects.
###### Information: ACL Inheritance Scenarios

There are four ACL Asset Inheritance scenarios:
- item <-> section(s) <-> component <-> global config
  (ex: article <-> category <-> com_content <-> global config)
- item <-> component <-> global config
  (ex: category <-> com_content <-> global config)
- component <-> global config
  (ex: com_content <-> global config)
- global config (no asset inheritance)

Besides inheriting from the asset(s), ACL also inherits from parent User Group(s).

**Mantainers** please confirm this are the scenarios at play.
#### Testing Instructions
###### Pre-requisites

Use latest staging.
Clean install.
###### Test 1: Testing default ACL Inheritance on creating a new iitem

This tests creating a new item without explicit inhreting permissions
- Create a new category "Test ACL" category, don't change anything and save
- Edit the category
- Check category permissions are inhreited form com_config + com_content
- Check there is no explict "Allowed" or "Denied" permission.
- Change one permission to "Denied" and exit
- Now create a new article assign to "Test ACL" category, don't change anything and save
- Now edit the article
- Check the com_config + com_content + category permissions are all inhrited as they should.
- Check there is no explict "Allowed" or "Denied" permission.

Note: You can also check that before the patch the item(s) have explicity "Allowed" or "Denied" permission on create.
###### Test 2: Testing ACL asset/groups Inheritance

This test should be made in sequence.

1: This tests Global config child groups Inherit
- Go to Global config
- Change Public "Delete" to Denied
- Change Public "Edit" to Allowed
- Refresh the page.
- Change Registered "Edit state" to Denied
- Change Registered "Edit Own" to Allowed
- Refresh the page
- Check all Global config Permissions are all inhrited as they should.

2: This tests Global config -> Component Inherit
- Go to Articles (com_content) Config 
- Check the com_config permission are all inhrited as they should.
- Change Author "Edit Own" to Denied
- Refresh the page
- Check the child groups permissions are all inhrited as they should.

3: This tests Global config -> Component -> Item Inherit
- Create new category, name it "Test ACL" don't change anything and save
- Now edit the category
- Check the com_config + com_content permissions are all inhrited as they should.
- Change Editor "Edit" to Denied
- Refresh the page
- Check the child groups  permissions are all inhrited as they should.

4: This tests Global config -> Component -> Section -> Item Inherit
- Create a new article assign to "Test ACL" category, don't change anything and save
- Now edit the article
- Check the com_config + com_content + category permissions are all inhrited as they should.
- Change Manager "Edit" to Denied
- Refresh the page
- Check the child groups  permissions are all inhrited as they should.

You can and should make other tests to confirm all is ok.
###### Test 3: code review

Do a code review.
#### Still known issues (that will not be solved by this PR)

1: When creating a new item (not saving) it uses the calculated permissions from the component (item <-> component <-> global config).
But if we have a section too (item <-> section(s) <-> component <-> global config) this is not correct.
This is a **incorrect info bug**.

2: In "1:", it uses the component permission, but should use the calculated permissions for achild of the component/section.
This is a **incorrect info bug**.

3: If a component as a permission that doesn't exists in global config (ex: frontend editing in com_modules) by default we get "Not Allowed (Inherited)" when we should get "Not Allowed (Default)".
In resume, if doesn't exist in the parent asset it can't Inherit from it.
This is a **incorrect info bug**.

4: When changing a permission of an item that doesn't have a row in the asset table the row a new row is created.
This works fine for item <-> component <-> global config scenario and component <-> global config scenario.
But doesn't work properly for item <-> section(s) <-> component <-> global config scenario because a wrong parent asset id (the component) is stored.
This is a **incorrect ACL bug**, happens when there is no row in the asset table (ex: deleted or not created on update).

Note: This known issues are marked in comments as todo in the code.

@infograf768 @roland-d thanks for all your help on checking this.
And please check.
